File: frontend/src/types/prediction.ts
/** Types mirroring /model, /scenarios and /predictions. */

export type RiskLevel = 'low' | 'moderate' | 'elevated' | 'high'
export type DoseUnit = 'Gy' | 'mGy' | 'kGy' | 'rad'

export interface DoseUnitOption {
  unit: DoseUnit
  label: string
  to_gray: number
}

export interface ScenarioDefaults {
  dose: number
  dose_unit: DoseUnit
  exposure_duration_days: number
  temperature_kelvin: number
  mechanical_force_pn: number
}

export interface Scenario {
  scenario_id: string
  label: string
  summary: string
  radiation_class: string | null
  environment: string
  particle_group: string | null
  /**
   * False when the scenario is outside the model's trained vocabulary. Those
   * scenarios still simulate, but produce no ML estimate.
   */
  ml_supported: boolean
  ml_unsupported_reason?: string
  defaults: ScenarioDefaults
  notes: string
}

export interface ScenarioProvenance {
  status: string
  statement: string
  ml_coupling: string
  trained_scenarios: string[]
}

export interface ScenariosResponse {
  scenarios: Scenario[]
  dose_units: DoseUnitOption[]
  provenance: ScenarioProvenance
}

export interface HeldOutMetrics {
  split: string
  rows: number
  proteins: string[]
  mae: number
  rmse: number
  r2: number
}

export interface FeatureImportance {
  feature: string
  group: 'numeric' | 'categorical'
  importance: number
}

export interface ModelInfo {
  available: boolean
  status: 'ready' | 'degraded' | 'unavailable'
  model_name: string | null
  model_version: string
  scientific_status: string
  label_source: string | null
  scientifically_validated: boolean
  approved_use: string | null
  created_at_utc: string | null
  bundle_sha256: string | null
  sha256_verified: boolean
  schema_verified: boolean
  load_error: string | null
  warnings: string[]
  target_column: string | null
  feature_order: string[]
  numeric_features: string[]
  categorical_features: string[]
  categorical_vocabulary: Record<string, string[]>
  n_transformed_features: number | null
  supports_uncertainty: boolean
  uncertainty_note: string | null
  validation_metrics: HeldOutMetrics | null
  test_metrics: HeldOutMetrics | null
  train_proteins: string[]
  validation_proteins: string[]
  test_proteins: string[]
  replacement_requirement: string | null
  limitations: string[]
  top_feature_importances: FeatureImportance[]
}

export interface PredictionRequest {
  pdb_id?: string
  upload_id?: string
  chain_id: string
  scenario_id: string
  dose: number
  dose_unit: DoseUnit
  exposure_duration_days: number
  temperature_kelvin: number
  mechanical_force_pn: number
  random_seed: number
  top_n_residues: number
}

export interface ResiduePrediction {
  residue_id: string
  residue_type: string
  proxy_rank: number
  degradation_percent: number
  residue_sasa_norm: number
  residue_contact_count: number
  qualitative_susceptibility: string
  /** False means the one-hot block was all zeros: the estimate is unreliable. */
  residue_type_in_model_vocabulary: boolean
}

export interface PredictionAggregation {
  method: string
  risk_band_basis: string
  explanation: string
  n_residues_predicted: number
  n_residues_used_in_mean: number
  n_residues_excluded_unknown_type: number
  per_residue_min: number
  per_residue_max: number
  per_residue_std: number
  whole_chain_mean_note: string
  exclusion_note?: string
}

export interface PredictionResponse {
  prediction_id: string
  model_version: string
  model_status: string
  degradation_percent: number
  risk_level: RiskLevel
  /** Always null: the bundle exposes no calibrated uncertainty. */
  confidence: number | null
  warnings: string[]
  input_summary: {
    structure: Record<string, unknown>
    scenario: Record<string, unknown>
    used_by_model: Record<string, unknown>
    not_used_by_model: Record<string, unknown>
  }
  residue_predictions: ResiduePrediction[]
  aggregation: PredictionAggregation
  held_out_error: {
    supported: boolean
    note: string
    validation: HeldOutMetrics | null
    test: HeldOutMetrics | null
  }
  /**
   * Spread of the per-residue predictions behind the protein-level figure.
   * Not a confidence interval -- the bundle exposes no calibrated uncertainty,
   * so there is no coverage probability to report and no `confidence_level`.
   */
  prediction_dispersion?: {
    available: boolean
    note: string
    basis?: string
    sd?: number
    min_pct?: number
    max_pct?: number
    mean_pct?: number
    n_residues?: number
  }
  applicability_domain?: {
    classification: 'IN_VOCABULARY' | 'CAUTION' | 'OUT_OF_DOMAIN' | string
    basis: string
    reasons: string[]
    note: string
  }
  /**
   * Closest proteins in the measured dataset by scaled sequence-descriptor
   * distance. No similarity percentage: the distance has no principled mapping
   * onto one. Empty when the queried protein is not itself measured.
   */
  nearest_neighbors?: Array<{
    pdb_id: string
    distance: number
    baseline_stiffness_pnnm: number
    resolved: boolean
  }>
  /** Exact tree SHAP contributions for the top-ranked candidate residue. */
  local_feature_attributions?: Array<{
    feature: string
    value: string | number | null
    contribution: number
    direction: 'increase' | 'decrease'
  }>
  global_feature_importance?: Record<string, number>
  attribution_disclaimer?: string
}

File: backend/app/schemas/prediction.py
"""Prediction request/response schemas.

The request carries the full experiment configuration the user set in the UI,
but only a subset reaches the model. Fields the model cannot consume are marked
in their descriptions and echoed back in ``input_summary.not_used_by_model`` so
the interface can be explicit rather than implying a coupling that is not there.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class PredictionRequest(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    # --- Structure selection -----------------------------------------------
    pdb_id: str | None = Field(
        default=None,
        description="A four-character approved PDB id. Mutually exclusive with upload_id.",
    )
    upload_id: str | None = Field(
        default=None,
        description="Id returned by POST /proteins/upload. Mutually exclusive with pdb_id.",
    )
    chain_id: str = Field(default="A", max_length=4)

    # --- Consumed by the model ---------------------------------------------
    scenario_id: str = Field(description="Must be a scenario with ml_supported=true.")

    # --- Not consumed by the model (simulation / provenance only) ----------
    dose: float = Field(
        default=0.5, ge=0.0, le=1.0e6,
        description="NOT an ML model input. Recorded for provenance and used by the "
                    "simulation's radiation-damage parameterisation.",
    )
    dose_unit: Literal["Gy", "mGy", "kGy", "rad"] = "Gy"
    exposure_duration_days: float = Field(
        default=180.0, ge=0.0, le=100_000.0,
        description="NOT an ML model input. Provenance only.",
    )
    temperature_kelvin: float = Field(
        default=300.0, gt=0.0, le=1000.0,
        description="NOT an ML model input. Sets the OpenMM thermostat temperature.",
    )
    mechanical_force_pn: float = Field(
        default=0.0, ge=0.0, le=10_000.0,
        description="NOT an ML model input. Provenance only in this MVP; no external "
                    "pulling force is applied by the Rapid Demo engine.",
    )
    random_seed: int = Field(default=42, ge=0, le=2**31 - 1)
    top_n_residues: int = Field(
        default=10, ge=1, le=50,
        description="How many ranked candidate residues to score. The model was "
                    "trained on the top 10; larger values extrapolate to less "
                    "susceptible residues than it ever saw.",
    )

    @model_validator(mode="after")
    def _exactly_one_structure(self) -> PredictionRequest:
        if bool(self.pdb_id) == bool(self.upload_id):
            raise ValueError(
                "Provide exactly one of 'pdb_id' (approved protein) or 'upload_id' "
                "(uploaded structure)."
            )
        return self


class ResiduePredictionOut(BaseModel):
    residue_id: str
    residue_type: str
    proxy_rank: float
    degradation_percent: float
    residue_sasa_norm: float
    residue_contact_count: float
    qualitative_susceptibility: str
    residue_type_in_model_vocabulary: bool


class PredictionResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    prediction_id: str
    model_version: str
    model_status: str = Field(
        description="Scientific status of the bundle, e.g. MOCK_PUBLIC_DATA_BOOTSTRAP."
    )
    degradation_percent: float = Field(
        description="Protein-level ML degradation estimate: the mean over the ranked "
                    "candidate residues. See 'aggregation' for how it is built."
    )
    risk_level: Literal["low", "moderate", "elevated", "high"]
    confidence: float | None = Field(
        default=None,
        description="Always null: the bundle exposes no calibrated uncertainty. "
                    "See held_out_error for retrospective dataset-level metrics.",
    )
    warnings: list[str]
    input_summary: dict[str, Any]
    residue_predictions: list[ResiduePredictionOut] = Field(default_factory=list)
    aggregation: dict[str, Any] = Field(default_factory=dict)
    held_out_error: dict[str, Any] = Field(default_factory=dict)
    prediction_dispersion: dict[str, Any] = Field(
        default_factory=dict,
        description=(
            "Spread of the per-residue predictions behind the protein-level "
            "figure. Not a confidence interval: the bundle exposes no "
            "calibrated uncertainty, so no coverage probability applies."
        ),
    )
    applicability_domain: dict[str, Any] = Field(
        default_factory=dict,
        description="Residue-vocabulary coverage of the input. Carries no numeric score.",
    )
    nearest_neighbors: list[dict[str, Any]] = Field(
        default_factory=list,
        description=(
            "Closest proteins in the measured dataset by scaled sequence-descriptor "
            "distance. Empty when the queried protein is not itself measured."
        ),
    )
    local_feature_attributions: list[dict[str, Any]] = Field(
        default_factory=list,
        description="Exact tree SHAP contributions for the top-ranked candidate residue.",
    )
    global_feature_importance: dict[str, float] = Field(
        default_factory=dict,
        description="The fitted estimator's own feature importances, largest first.",
    )
    attribution_disclaimer: str | None = None


class ModelInfoResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    available: bool
    status: str = Field(description="ready | degraded | unavailable")
    model_name: str | None = None
    model_version: str
    scientific_status: str
    label_source: str | None = None
    scientifically_validated: bool = False
    approved_use: str | None = None
    created_at_utc: str | None = None
    bundle_sha256: str | None = None
    sha256_verified: bool = False
    schema_verified: bool = False
    load_error: str | None = None
    warnings: list[str] = Field(default_factory=list)
    target_column: str | None = None
    feature_order: list[str] = Field(default_factory=list)
    numeric_features: list[str] = Field(default_factory=list)
    categorical_features: list[str] = Field(default_factory=list)
    categorical_vocabulary: dict[str, list[str]] = Field(default_factory=dict)
    n_transformed_features: int | None = None
    supports_uncertainty: bool = False
    uncertainty_note: str | None = None
    validation_metrics: dict[str, Any] | None = None
    test_metrics: dict[str, Any] | None = None
    train_proteins: list[str] = Field(default_factory=list)
    validation_proteins: list[str] = Field(default_factory=list)
    test_proteins: list[str] = Field(default_factory=list)
    replacement_requirement: str | None = None
    limitations: list[str] = Field(default_factory=list)
    top_feature_importances: list[dict[str, Any]] = Field(default_factory=list)
